[MERGE][IMP] web_editor: Implement collaborative for Knowledge

abd-msyukyu-odoo opened the pull request:
# Editor changes to enable the collaborative mode in Knowledge
Here are the necessary changes to make it possible to enable the collaborative 
mode in Knowledge. Knowledge commands introduce a new kind of editor blocks:
`Behaviors`, which are OWL `Component` which need to be mounted. This mounting
is asynchronous and this has to be supported, while not all nodes are observed 
by the editor, meaning that parts of a `Behavior` can be collaborative while 
others may not.

The following is a brief introduction for every commit in this PR, check them
out individually for more detail.

### [FIX] web_editor: notify current step when joining a collaborator
- Inserting an embedded view from a module in an automated process happens as
  soon as the first collaborator connect. This commit notifies the step for the
  embedded view during the following collaborators connections (to be sure they
  actually received it).

### [IMP] web_editor: fully implement oeProtected and oeTransientContent
- Those attributes will allow a fine control on which nodes are NOT observed by
  the editor (`oeProtected="true"`), and which nodes are NOT saved/serialized
  for collaboration by the editor (`oeTransientContent="true"`). It can be a
  better choice than using `observerUnactive` which stops the observer for all
  nodes of the editor, in case a node should never be observed.

### [IMP] web_editor: prevent 'insert' command unwrapping in root
- Modify the 'insert' command of the editor so that it does not unwrap nodes
  when the selection is directly inside the editable ('root'). Unwrapping a
  paragraph in that case would put text nodes as direct childs of the editable,
  which is not desirable with the current implementation.

### [IMP] web_editor: add a hook for historyResetFromSteps
- When joining an `html_field` in collaborative mode, there is no way to know if
  there are connected collaborators before actually making said connection. This
  is a problem when we want to automate the insertion of an embedded view from
  another module in Knowledge. With the hook, it is now possible to execute a
  specific action as soon as a collaborator is joined.

### [IMP] web_editor: force load collaborative on form view
- The collaborative mode loads lazily with the focus by default. In Knowledge it
  should start as soon as possible, therefore this commit adds an option to do
  so.

### [IMP] web_editor: make mocha tests odoo modules
- Since `web_editor` tests were not odoo modules, it was not possible to import
  dependencies from i.e. `web`. This commit addresses that.

### [IMP] web_editor: buffer external steps during Component rendering
- In a collaboration, when applying a step requiring a custom OWL `mount`
  (which is asynchronous), further external steps should be buffered while the
  rendering is done, in case those following external steps should concern
  "rendered" nodes which are not yet present in the editor. Once all nodes are
  rendered and present in the editor, further steps can be applied safely.

See odoo/enterprise#33483
See odoo/upgrade#4384

robodoo commented:
[Pull request status dashboard](https://mergebot.odoo.com/odoo/odoo/pull/104680).

abd-msyukyu-odoo commented on addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:
```diff
@@ -664,33 +695,7 @@ const Wysiwyg = Widget.extend({
+                const firstClientId = this.getFirstClientId();
                 if (firstClientId !== this._currentClientId) {
                     const historySteps = await this.ptp.requestClient(firstClientId, 'get_history_from_snapshot', undefined, { transport: 'rtc' });
                     this.odooEditor.historyResetFromSteps(historySteps);
```
ici -> inserer la vue du service (embedded view)

Goaman commented on addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:
```diff
@@ -553,6 +555,7 @@ const Wysiwyg = Widget.extend({
                                     if (remoteSelection) {
                                         this.odooEditor.onExternalMultiselectionUpdate(remoteSelection);
                                     }
+                                    this.underwentCollabSync.resolve(true);
```
`this._onSync();`

Goaman commented on addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:
```diff
@@ -2552,6 +2561,9 @@ const Wysiwyg = Widget.extend({
     },
     _stopPeerToPeer: function () {
         this.ptp && this.ptp.stop();
+        if (this.underwentCollabSync) {
```
`this._onSync = undefined;`

dmo-odoo commented on addons/web_editor/static/src/js/editor/odoo-editor/test/spec/insertHTML.test.js:
```diff
@@ -90,6 +93,20 @@ describe('insert HTML', () => {
                 contentAfter: '<pre>abcdef[]<br>ghi</pre>',
             });
         });
+        it('should unpack elements from a lone paragraph (Knowledge check)', async () => {
```
It's kind of weird to explicitely reference knowledge in an editor community test.

dmo-odoo commented on addons/web_editor/static/src/js/editor/odoo-editor/test/spec/insertHTML.test.js:
```diff
@@ -90,6 +93,20 @@ describe('insert HTML', () => {
+                },
+                contentAfter: '<p><br></p><div>content</div><p><br>[]</p>',
+            });
+        });
```
That is in fact a bug, I think we discussed about it already. Maybe I did not create a task ? Or we just haven't done it yet ? I need to check.

dmo-odoo commented on addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:
```diff
@@ -2552,6 +2554,7 @@ const Wysiwyg = Widget.extend({
     },
     _stopPeerToPeer: function () {
         this.ptp && this.ptp.stop();
+        this._onHistoryResetFromSteps = () => {};
```
Wait, those `_onHistoryResetFromSteps` changes actually have an effect on the unpacking ? I'm confused.

abd-msyukyu-odoo commented on addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:
```diff
@@ -953,11 +955,25 @@ export class OdooEditor extends EventTarget {
+                        break;
+                    case "false":
+                        if (record.target === closestProtectedElement && record.type === "attributes" &&
+                            isProtected(closestProtectedElement.parentElement)) {
```
-> use Prettier syntax

abd-msyukyu-odoo commented on addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:
```diff
@@ -953,11 +955,25 @@ export class OdooEditor extends EventTarget {
+                    case "blackbox":
+                    case "true":
+                        if (record.type !== "attributes") {
+                            this.idSet(record.target);
```
go through mutations to apply idSet on every mutations

abd-msyukyu-odoo commented on addons/web_editor/static/src/js/backend/html_field.js:
```diff
@@ -659,6 +659,14 @@ export const htmlField = {
         if ('collaborative' in options) {
             wysiwygOptions.collaborative = options.collaborative;
+            switch(options.collaborative_trigger) {
+                case 'start':  // join the peerToPeer connection immediately
```
Join

abd-msyukyu-odoo commented on addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:
```diff
@@ -1518,19 +1525,34 @@ export class OdooEditor extends EventTarget {
-            this._historyAddExternalStep(newStep);
-        }
+            for (const newStep of newSteps) {
+                await this._historyAddExternalStep(newStep);
```
should not be async and no need to await

dmo-odoo commented on addons/web_editor/static/src/js/editor/odoo-editor/test/spec/insert.test.js:
```diff
@@ -90,6 +93,21 @@ describe('insert HTML', () => {
+            // unpacking content of a paragraph or from the first and last
+            // inserted elements (in case there are more than one). Don't forget
+            // to update the appendBehaviorBlueprint method in Knowledge
+            // wysiwyg.js at that point.
```
It would probably be better to fix this rather than crystalize the wrong behavior inside a test. We could move the two conditions that test `container.childElementCount` from `insert` into `_onPaste`. I had a quick look and I don't think commands other than insert _itself_ rely on this behavior. The only dubious case I found that needs extra checks is the `/sign` command.

Goaman commented on addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:
```diff
@@ -765,6 +768,21 @@ export class OdooEditor extends EventTarget {
         return this._idToNodeMap.get(id);
     }
 
+    /**
```
remove this commit

Goaman commented on addons/web_editor/static/src/js/editor/odoo-editor/src/OdooEditor.js:
```diff
@@ -1013,6 +1019,25 @@ export class OdooEditor extends EventTarget {
+     *          rendering to do. The editor will buffer new external
+     *          steps (collaborative) until the promise is resolved.
+     */
+    checkForExtraRendering() {
```
postProcessExternalSteps

abd-msyukyu-odoo commented:
@dmo-odoo Here is the new state of the 'insert' command:
https://github.com/odoo/odoo/pull/104680/commits/00f16b8640671c9c8ef55dab39fa3235ccf60349
what do you think?

dmo-odoo commented:
For followup purposes, we just decided with @abd-msyukyu-odoo that we would make a special case for inserting at the root instead because I do not think the developper should handle the unwrapping manually. I think it is correct that the insert command handles the unwrapping when it is necessary. The catch is to nail the "when it is necessary" part right, hence the suggested change.

dmo-odoo commented on addons/web_editor/static/src/js/wysiwyg/wysiwyg.js:
```diff
@@ -694,6 +694,9 @@ const Wysiwyg = Widget.extend({
         };
+        if (this.options.checksforExtraRendering) {
+            editorCollaborationOptions.checksforExtraRendering = this.options.checksforExtraRendering;
+        }
```
?

abd-msyukyu-odoo commented:
@dmo-odoo Thanks a lot for proofreading this again, it seems I didn't have "les yeux en face des trous" when we discussed the renaming of `postProcessExternalSteps`

I added a test for that feature (in the commit "buffering external steps") and reworked collaborative tests so steps can be asynchronous. Also added a way for the test to provide options for the `OdooEditor` instance used in the test.

For the `_onHistoryResetFromSteps` hook, as always, I don't mind the form, but I need the feature, be it an option, a hook, a promise, ... ;p

dmo-odoo commented:
@robodoo r+ Most of the changes are in `web_editor`. I let the other people responsible for Knowledge to handle the rebase method and the other linked PRs.

dbeguin commented:
@xmo-odoo the merge method will defined by us. so that was intended :) (for that part anyway)

dmo-odoo commented:
@robodoo r+

Trying to find another way to write my novels without using two messages. This one uses new lines, hoping the mergebot will not crash this time. 🤞

dbeguin commented:
@robodoo rebase-merge

robodoo commented:
Merge method set to rebase and merge, using the PR as merge commit message.

robodoo commented:
@abd-msyukyu-odoo @dmo-odoo linked pull request(s) odoo/enterprise#33483 not ready. Linked PRs are not staged until all of them are ready.

robodoo commented:
@abd-msyukyu-odoo @dmo-odoo linked pull request(s) odoo/upgrade#4384 not ready. Linked PRs are not staged until all of them are ready.
